src/lib/heimgard.ts

```diff
@@ -0,0 +1,214 @@
+            const lockChangeRequester = utils.getFromLookup(msg.data.opereventsrc, lockChangeSource);
+            const isLocked = lockStateCode === 1;
+
+            let {isInteriorLocked, isExteriorLocked} = meta.state;
+
+            // Autolocked by device
+            if (lockChangeRequester === "self") {
+                const targetLock = identifyLockStateFromHistory(meta);
+                if (targetLock === LockSide.Interior) {
+                    isInteriorLocked = isLocked;
+                    result.inner_lock_state = "locked";
+                } else if (targetLock === LockSide.Exterior) {
+                    isExteriorLocked = isLocked;
+                    result.lock_state = "locked";
+                    result.state = lockStateCode === 1 ? "LOCK" : "UNLOCK";
+                    meta.state.state = result.state;
+                } else {
+                    void lock.convertSet(msg.endpoint, "state", "LOCK", null);
+                    return;
+                }
+            } else {
+                if (lockChangeRequester === "function_key") {
+                    isInteriorLocked = isLocked;
+                    result.inner_lock_state = isLocked ? "locked" : "unlocked";
+                } else {
+                    if (!isLocked) {
+                        result.last_unlock_source = lockChangeRequester;
+                        if (lockChangeRequester === "function_key" || lockChangeRequester === "remote") {
+                            result.last_unlock_by_user = "N/A";
+                        } else {
+                            result.last_unlock_by_user = msg.data.userid;
+                        }
+                    } else {
+                        // If locked
+                        result.last_lock_source = lockChangeRequester;
+                        result.last_lock_by_user = msg.data.userid;
+                    }
+
+                    isExteriorLocked = isLocked;
+                    const known_lockstates = {
+                        0: "unknown_lock_failure",
+                        1: "locked",
+                        2: "unlocked",
+                    };
+                    result.lock_state = utils.getFromLookup(lockStateCode, known_lockstates, "unknown_lock_failure");
+                    result.state = isLocked ? "LOCK" : "UNLOCK";
+                    meta.state.state = result.state;
+                }
+            }
+
+            const newHistoryEntry: LockStateHistory = {
+                time: Date.now(),
+                interiorLockState: isInteriorLocked ? "locked" : "unlocked",
+                exteriorLockState: isExteriorLocked ? "locked" : "unlocked",
+            };
+            const historyCache = (meta.state.history as LockStateHistory[]) ?? [];
+            historyCache.push(newHistoryEntry);
+            // Keeping only the last 5 entries
+            meta.state.history = historyCache.slice(-5);
+
+            meta.state.isInteriorLocked = isInteriorLocked;
```

Comment:
Please don't put values in `meta.state` here, as things will get published to MQTT later. Instead use the `globalStore` (search this repo for examples)